bin/consequence_mapping/consequence_mapping.py

```diff
@@ -0,0 +1,154 @@
+            consequences are sorted by distance from the gene and the closest one is chosen. In case of a tie,
+            consequence is selected at random. If this flag is not specified, all consequences for each variant will
+            be returned.
+        results_by_variant: a dict to write the results into.
+    """
+    for result in vep_results:
+        variant_identifier = result['input']
+        results_by_variant.setdefault(variant_identifier, [])
+        consequences = result.get('transcript_consequences', [])
+
+        # Keep only consequences with the allowed biotypes; if there are none, skip this variant
+        consequences = [c for c in consequences if c['biotype'] in acceptable_biotypes]
+        if not consequences:
+            continue
+
+        # Flatten the list of consequence terms and find the most severe one
+        all_consequence_terms = [term for c in consequences for term in c['consequence_terms']]
+        all_consequence_terms.sort(key=lambda term: consequence_term_severity_rank[term])
+        most_severe_consequence_term = all_consequence_terms[0]
+
+        # Keep only consequences which include the most severe consequence term; sort by increasing order of distance
+        consequences = [c for c in consequences if most_severe_consequence_term in c['consequence_terms']]
+        consequences.sort(key=lambda consequence: abs(consequence.get('distance', 0)))
+
+        # If mandated by a flag, keep only one (least distant) consequence
+        if only_closest:
+            consequences = [consequences[0]]
+
+        # Return a subset of fields (required for output) of filtered consequences
+        results_by_variant[variant_identifier].extend([
+            (variant_identifier, c['gene_id'], c['gene_symbol'], most_severe_consequence_term, c.get('distance', 0))
+            for c in consequences
+        ])
+
+
+def get_variants_without_consequences(results_by_variant):
+    """Returns a list of variants for which no consequences were found."""
+    return sorted({
+        variant_id
+        for variant_id, list_of_consequences in results_by_variant.items()
+        if len(list_of_consequences) == 0
+    })
+
+
+def process_variants(variants):
+    """Given a list of variant IDs, return a list of consequence types (each including Ensembl gene name & ID and a
+    functional consequence code) for a given variant."""
+
+    # First, we query VEP with default parameters, looking for variants affecting protein coding and miRNA transcripts
+    # up to a standard distance (5000 nucleotides either way, which is default for VEP) from the variant.
+    results_by_variant = {}
+    vep_results = query_vep(variants=variants, search_distance=5000)
+    extract_consequences(vep_results=vep_results, acceptable_biotypes={'protein_coding', 'miRNA'},
+                         only_closest=False, results_by_variant=results_by_variant)
```

Comment:
Judgement call: Just from a programmer POV, I am a bit antsy about functions that mutate the variables that are passed in. Could we make **extract_consequences** explicitly return the modified dictionary so that we can assign the function call result to results_by_variant again? That makes it easier to figure out the places where the variable gets modified...

Reply:
This is a very good point! Implemented the way you suggested